### ISSUE TITLE
Counts do not match taginfo

### ISSUE THREAD

**Issue by brawer** [astro/rust-osm-pbf-iter#10]
When iterating over the OSM planet dump with [count.rs](https://github.com/astro/rust-osm-pbf-iter/blob/master/src/bin/count.rs), the counts seem too low. Does the library skip some blobs?

```
$ aria2c https://planet.openstreetmap.org/pbf/planet-231225.osm.pbf.torrent
$ target/release/count planet-231225.osm.pbf 
Open planet-231225.osm.pbf
Processed 74023 MB in 93.18 seconds (794.38 MB/s)
planet-231225.osm.pbf - 445152000 nodes, 117960000 ways, 73800 relations
```

Expected same counts as reported by [taginfo](https://taginfo.openstreetmap.org/reports/database_statistics) or [OSM stats](https://planet.openstreetmap.org/statistics/data_stats.html): 8848241030 nodes, 988662259 ways, 11658204 relations

**Comment by astro** [comment]
Is there any other tool that we can feed the same .pbf file for direct comparison, perhaps with a smaller dump?

Is it possible that taginfo/OSM stats count deleted nodes?

**Comment by astro** [comment]
Ok, osmpbfreader-rs' count example reports:

```
230280468 nodes, mean coord: 1191.5522999446323, 260.8464536548693.
988662259 ways, mean |nodes|: 10.608699534670919
11658204 relations, mean |references|: 11.262824359566876
```

Our numbers are oddly even...

**Comment by brawer** [comment]
Possibly related: When iterating over the OSM planet of 2026-01-19, quite a lot of nodes are not yielded. For example, [node 1221970421](https://www.openstreetmap.org/node/1221970421) never gets yielded.

Interestingly, according to `osm-pbf-iter`, each blob in the PBF file contains ~20K primitives. However, between the blobs, there seems to be a gap of around 200K IDs, which smells suspicious. For example, when iterating over data blob 3520 of the 2026-01-19 planet file, this library yields nodes 1_221_906_996 to 1_221_927_261. When iterating over the successor blob 3521, the iterator yields nodes 1_222_157_290 to 1_222_176_700. Which is ~20K ids for each blob, with a 230K gap in between. The missing node 1_221_970_421 falls into this gap.

This makes me speculate if we’re perhaps only iterating over the _first_ PrimitiveGroup in a PrimitiveBlock, instead of all of them? If that was the case, it could be an explanation for the oddly even numbers we found earlier (“445152000 nodes, 117960000 ways, 73800 relations”): if an encoder stuffs 100 or 1000 primitives into the same PrimitiveGroup, with the remainder going into the last group in the block or file, we would see multiples of 100 or 1000 if some bug makes us fail to decode all groups.

**Comment by brawer** [comment]
Indeed, this problem still exists, and looks like we’re not the only ones having it. Counting OpenStreetMap features in the 2026-01-19 planet file, it seems that only osmpbf is getting it right:

| Library  | Nodes |    Ways |   Relations | Runtime | Peak RAM |
|-------- | --------: | -----: | -------: | -----: | -----: |
| [osmpbf](https://github.com/b-r-u/osmpbf/blob/main/examples/count.rs)      | 10_375_842_718 | 1_157_027_150 | 14_026_055 | 432 s | 189 MB |
| [osmpbfreader-rs](https://github.com/TeXitoi/osmpbfreader-rs/blob/master/examples/count.rs)  |   1_785_908_126 | 1_157_027_150  | 14_026_055 | 1351 s | 1397 MB |
| [rust-osm-pbf-iter](https://github.com/astro/rust-osm-pbf-iter/blob/master/src/bin/count.rs) | 522_000_000 | 139_184_000 | 89_800 | 125 s | 80 MB |

Runtime and Peak RAM measured with `/usr/bin/time -lh` on a 2018 MacBook Pro with 6 Intel CPU cores, using the linked code examples of various Rust libraries, taken from their respective main branches on GitHub as of 2026-03-03.

**Comment by brawer** [comment]
Looks like we’re indeed only parsing the very first PrimitiveGroup in a PrimitiveBlock, whereas `osmpbf` parses them all.

I’ve been able to construct a small file for testing this, using a little [ad-hoc script](https://gist.github.com/brawer/40e316eef456f23521f088e6e70db3c3). The test file contains a single data blob with two primitive groups; the first with one node, the second with two nodes. Current `rust-osm-pbf-iter` finds just 1 node in this file, while `osmpbf` finds 3.

I’ll add this test file to `tests/data`, and write a test case that reads it.

**Comment by brawer** [comment]
OK, I have a fix ready. After the change, we get the same counts as the osmpbf library, but in half the time:

| Library  | Nodes |    Ways |   Relations | Runtime | Peak RAM |
|-------- | --------: | -----: | -------: | -----: | -----: |
| [osmpbf](https://github.com/b-r-u/osmpbf/blob/main/examples/count.rs)      | 10_375_842_718 | 1_157_027_150 | 14_026_055 | 432 s | 189 MB |
| [osmpbfreader-rs](https://github.com/TeXitoi/osmpbfreader-rs/blob/master/examples/count.rs)  |   1_785_908_126 | 1_157_027_150  | 14_026_055 | 1351 s | 1397 MB |
| [rust-osm-pbf-iter](https://github.com/astro/rust-osm-pbf-iter/blob/master/src/bin/count.rs) | 10_375_842_718 | 1_157_027_150 | 14_026_055 | 219 s | 129 MB |
